driver-4/src/main/java/com/azure/cosmos/cassandra/CosmosLoadBalancingPolicy.java

```diff
@@ -0,0 +1,632 @@
+        }
+
+        final Function<Request, Queue<Node>> function = this.getNodes;
+        final Queue<Node> nodes = function.apply(request);
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("newQueryPlan -> returns({}), {}", Json.toJson(nodes), this);
+        }
+
+        return nodes;
+    }
+
+    @Override
+    public void onAdd(@NonNull final Node node) {
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("onAdd({})", Json.toJson(node));
+        }
+
+        if (this.preferredRegions.contains(node.getDatacenter())) {
+            this.distanceReporter.setDistance(node, NodeDistance.LOCAL);
+        } else {
+            this.distanceReporter.setDistance(node, NodeDistance.REMOTE);
+        }
+
+        this.nodesForReading.add(node);
+
+        if (!this.multiRegionWrites) {
+            if (this.getContactPointsOrException().contains(node)) {
+                this.nodesForWriting.add(node);
+            }
+        }
+
+        LOG.debug("onAdd -> returns(void), {}", this);
+    }
+
+    @Override
+    public void onDown(@NonNull final Node node) {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("onDown({})", Json.toJson(node));
+        }
+    }
+
+    @Override
+    public void onRemove(@NonNull final Node node) {
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Removing node: onRemove({})", Json.toJson(node));
+        }
+
+        this.nodesForReading.remove(node);
+
+        if (!this.multiRegionWrites) {
+            this.nodesForWriting.remove(node);
+        }
+
+        LOG.debug("onRemove -> returns(void), {}", this);
+
+        if (this.multiRegionWrites) {
+            if (this.nodesForReading.isEmpty()) {
+                LOG.warn("All nodes have now been removed: {}", this);
+            }
+        } else {
+            if (this.nodesForReading.isEmpty()) {
+                LOG.warn("All nodes for reading have now been removed: {}", this);
+            }
+            if (this.nodesForWriting.isEmpty()) {
+                LOG.warn("All nodes for writing have now been removed: {}", this);
+            }
+        }
+    }
+
+    @Override
+    public void onUp(@NonNull final Node node) {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("onUp({})", Json.toJson(node));
+        }
+    }
+
+    @Override
+    public String toString() {
+        return Json.toString(this);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Nullable
+    private Set<Node> getContactPoints() {
+        return (Set<Node>) (Set<?>) this.driverContext.getMetadataManager().getContactPoints();
+    }
+
+    @NonNull
+    private Set<Node> getContactPointsOrException() {
+        return requireNonNull(this.getContactPoints(), "expected non-null contactPoints");
+    }
+
+    // endregion
+
+    // region Privates
+
+    private Queue<Node> doGetNodes(final Request request) {
+        return new ConcurrentLinkedQueue<>(isReadRequest(request) ? this.nodesForReading : this.nodesForWriting);
```

Comment:
we might need to optimize this to avoid instantiating a new object, to avoid perf increase on client side.